src/osd/scrubber/scrub_machine.cc

```diff
@@ -1149,6 +1149,14 @@ sc::result ReplicaActiveOp::react(const StartReplica&)
   return transit<ReplicaActiveOp>();
 }
 
+sc::result ReplicaActiveOp::react(const ReplicaRelease& ev)
+{
+  dout(10) << "ReplicaActiveOp::react(const ReplicaRelease&)" << dendl;
+  post_event(ev);
+  return transit<sc::shallow_history<ReplicaReserved>>();
```

Comment:
It seems like this should be a simple transit to ReplicaActive.  What prevents that from working?

Reply:
It won't work because of the way the history mechanism is used in statechart.
The desire is to return to either "reserved" or "unreserved" substates.
(Actually - I forgot about that in my first version of this fix, and it crushed nicely)

Reply:
Can you expand on that?

Reply:
Sure. Take a look at https://github.com/ronen-fr/schistory.
This repository contains my attempts to get boost::statechart history right.

What I've learnt is that, if I need to return to the specific sub-state
of ReplicaIdle - i.e., either ReplicaReserved or ReplicaUnreserved,
I'll need to:

(all this already exists in the code before the PR):

- mark ReplicaActive & ReplicaIdle as having shallow history (l. 834)
- clear the history when becoming irrelevant (e.g. l. 763)
- when requiring a transition back in ReplicaIdle that must
return to the appropriate sub-state - use the 
following  command:
 `transit<sc::shallow_history<ReplicaReserved>>();`

 Some of this was counter-intuitive for me - but proved to work when tested in 'schistory'.

Reply:
I...missed that you started using the history mechanism in https://github.com/ceph/ceph/pull/55131.  I tried reading through the boost documentation, but it really doesn't give any information on how it works or what it does beyond mentioning the templates and types (unless I missed something?).

```
 *    - uses 'shallow history', so that when returning from ReplicaActiveOp, we
 *       return to where we were - either reserved by our primary, or unreserved.
```

seems to be the only comment mentioning it.  I confess -- I simply do not understand this mechanism and am not comfortable reviewing this PR until I do.  Nothing in ReplicaActiveOp jumps out to me as making use of history and I don't understand how the ```transit<sc::shallow_history<ReplicaReserved>>()``` calls work.  The repository you link above really isn't all that helpful as I don't really have time at the moment to experiment with it -- that's going to be an ongoing problem for anyone new who wants to review or debug code here.

One option would be to add enough information to the comment above to allow new readers/reviewers to understand how the feature works and how it fits into your design.

My larger concern, however, is that it's just not clear to me why history would be helpful.  The most obvious way (to me, at least) to model this would be two orthogonal sets of substates:
1. reserved/unreserved
2. state sequence for handling a chunk request

reserved/unreserved doesn't even need to be a fully fledged set of states -- it could just be a boolean on ReplicaActive.  Returning to a previous state just doesn't seem like a natural solution to the problem of tracking whether we happen to be holding a reservation.  It seems like the intention here is for a Release to release the reservation and halt any in-progress scan -- it seems like exiting our current states and entering ReplicaActive should handle that naturally and it strikes me as odd that it seems not to. 

If there's an existing explanation/documentation I should be reading, I'm perfectly happy to do that and give it another try, though.

Reply:
@athanatos ,
The boost::statechart documentation regarding 'history' is indeed extremely scarce. There is an implicit assumption there, that the details of 'shallow' vs. 'deep' history mechanism - exactly as implemented in statechart - are common knowledge. I don't think that's true.

I used some external resources, including:

https://anylogic.help/anylogic/statecharts/history-state.html

https://www.ibm.com/support/pages/general-information-deep-and-shallow-history-connector-rhapsody


My preferred solution:
After my experience with the boost implementation, I think the best (simplest to maintain) solution is to have a reserved/unreserved flag in the replica based state
(i.e. - implementing a shallow-history-like mechanism for this specific flag ourselves). There are two options:

- The cleanest (and my pref) is to have a transitory initial state in ReplicaIdle, that checks the flag and immediately transitions into Reserved or Unreserved.
  This involves minimal changes to the code.

- another solution is not adding a base state, and instead check the flag on each reaction (using it as an explicit  'guard'). That would result in a clumsier code.

Note the current context:

The bug exists in Squid, and must be fixed. It is also creating multiple failures in teuthology tests today.
I would suggest the following steps:
- merge this PR (after I add more context to the commit description, as requested) into Main;
- backport the fix into Squid;
- in Main only - I can prepare a new implementation based on the above in a couple of days.

Reply:
@athanatos - I am now testing a PR (https://github.com/ceph/ceph/pull/56192) that removes the
history mechanism from the scrub FSM, replacing it with a safe internal transition.

Reply:
I'm fine with merging this to take care of the bug for now.  I want to backport the history removal PR once ready to squid -- it hasn't been released yet and it's worth doing to avoid having a stable version with this structure.

I don't think you need both the boolean and Reserved/Unreserved states.  The only purpose of the boolean is so that upon exit we know to release the reservation.  ReplicaActive would have the reactions for reservation, reservation complete, and release.  The other states/behavior do not care about the value of that boolean at all.